bowling: add exercise

leenipper opened the pull request:
Example solution modeled somewhat on http://butunclebob.com/ArticleS.UncleBob.TheBowlingGameKata

petertseng commented on exercises/bowling/bowling_test.go:
```diff
@@ -0,0 +1,68 @@
+func applyPreviousRolls(g *Game, rolls []int) error {
+	for _, pins := range rolls {
+		if err := g.Roll(pins); err != nil {
+			return err
```
It could be possible to have a more helpful error message when `applyPreviousRolls` errors if this would return:

* the index of the erroring roll
* the number of pins in the erroring roll

I think that would help a student debug if there is a problem applying previous rolls.

leenipper replied:
That had crossed my mind.  It is a good idea. I'll put that in.

petertseng commented on exercises/bowling/bowling_test.go:
```diff
@@ -0,0 +1,68 @@
+		if tc.valid {
+			if err != nil {
+				t.Fatalf("%s : Score() after Previous Rolls: %#v expected %d, got error %s",
+					tc.description, tc.previousRolls, tc.score, err)
```
Potential to apply #275 here (ensuring the second return value is of type `error`). I'm not sure it's necessary, but we could either add a `var _ error = err` or call `err.Error()` explicitly? Or should we not bother applying #275 here?

I should document #275 somewhere since everyone keeps forgetting it, maybe it's not really necessary if students by this point already understand that the last return value should be of type `error`?

leenipper replied:
Ok, got it.  Easy to do.

petertseng commented on exercises/bowling/example.go:
```diff
@@ -0,0 +1,126 @@
+// An error is possible depending on pin value and previous rolls.
+func (g *Game) Roll(pins int) error {
+	// Validate pin count on roll.
+	if pins > 10 {
```
a lot of magic numbers in here, could they become consts such as `pinsPerFrame`? I know they'll never change so it we don't get the advantage of "easily be able to change the constant if we need to", but we do get the advantage of "what exactly does this number mean and why is it 10?" - and this is doubly important because there are multiple 10s in this code and some of them "number of frames in a game" and some of them are "number of pins in a frame"

leenipper replied:
Ah, I had overlooked that because bowling has the "fixed" constants.  Good point about 10 frames vs. 10 pins.  Will fix it up.

leenipper commented:
> returned errors contain exactly the errorString

I had not intended to require that the student error matches errorString, but I understand why it appears that way. Since the canonical data file had the error texts, I thought it was helpful to carry them into the output.

> just the explanation for why there would be an error

This is my preference, to alter the test fail output a bit to make sure it isn't taken as a required error string. I could even change the test case member to explainText for a lighter impression of its use.
I think leaving the error specifics up to the student allows more flexibility and creativity in solutions.
Other thoughts, anyone ?

petertseng commented on exercises/bowling/bowling_test.go:
```diff
@@ -39,8 +47,8 @@ func TestScore(t *testing.T) {
 		} else if err == nil {
-			t.Fatalf("%s : Score() after Previous Rolls: %#v expected error \"%s\", got score %d",
-				tc.description, tc.previousRolls, tc.errorString, score)
+			t.Fatalf("FAIL: %s : Score() after Previous Rolls: %#v expected an error, got score %d\n\tExplaination: %s",
```
spelling: Explaination -> Explanation

It does mess me up sometimes that when you explain something, you give an explanation... the i disappeared

leenipper replied:
oops. ok.

petertseng commented on exercises/bowling/bowling_test.go:
```diff
@@ -49,19 +57,20 @@ func TestScore(t *testing.T) {
 		} else if !tc.valid && err == nil {
-			t.Fatalf("%s : Roll(%d) after Previous Rolls: %#v did not get expected error \"%s\".",
-				tc.description, tc.roll, tc.previousRolls, tc.errorString)
+			t.Fatalf("FAIL: %s : Roll(%d) after Previous Rolls: %#v expected an error.\n\tExplaination: %s",
```
same (spelling)

leenipper replied:
ok

petertseng commented on exercises/bowling/cases_test.go:
```diff
@@ -8,7 +8,7 @@ var scoreTestCases = []struct {
 	valid         bool   // true => no error, false => error expected
 	score         int    // when .valid == true, the expected score value
-	errorString   string // when .valid == false, error string detail
+	explainText   string // when .valid == false, error explaination text
```
same (spelling)

petertseng commented on exercises/bowling/cases_test.go:
```diff
@@ -160,7 +160,7 @@ var rollTestCases = []struct {
 	valid         bool   // true => no error, false => error expected
 	roll          int    // pin count for the test roll
-	errorString   string // when .valid == false, error string detail
+	explainText   string // when .valid == false, error explaination text
```
same (spelling)

leenipper replied:
ok

petertseng commented on exercises/bowling/example.go:
```diff
@@ -29,63 +37,62 @@ func NewGame() *Game {
 	g.nRolls++
-	if pins == 10 && g.completedFrames() < 9 {
-		// Frames 1 .. 9 can be strikes with no problems.
+	if pins == pinsPerFrame && g.completedFrames() < (framesPerGame-1) {
```
optional: the parentheses around `framesPerGame-1` aren't necessary

leenipper replied:
The parenthesis tend to highlight the expression.  But it is probably more idiomatic to leave them out.

petertseng commented on exercises/bowling/example.go:
```diff
@@ -29,63 +37,62 @@ func NewGame() *Game {
+		// Have counted normal max rolls on a frame.
+		if g.rawFrameScore(g.rFrameStart) > pinsPerFrame {
+			// Unless we have completed all but last frame, cannot count > pinsPerFrame.
+			if g.completedFrames() != (framesPerGame - 1) {
```
optional: the parentheses around `framesPerGame-1` aren't necessary

leenipper replied:
Interesting how gofmt will insert the spaces on _some_ instances of parenthesis use. I will go without them.

petertseng commented on exercises/bowling/example.go:
```diff
@@ -29,63 +37,62 @@ func NewGame() *Game {
 		}
-		if g.completedFrames() < 9 {
-			// Completed frames 1 .. 9 with 2 rolls.
+		if g.completedFrames() < (framesPerGame - 1) {
```
optional: the parentheses around `framesPerGame-1` aren't necessary

petertseng commented on exercises/bowling/example_gen.go:
```diff
@@ -53,14 +53,14 @@ func (c OneCase) Score() int {
 // determineExpected examines an .Expected interface{} object and determines
 // whether a test case is valid(bool), has a score field, and/or has an expected error,
-// returning valid, score, and error string text.
+// returning valid, score, and error explaination text.
```
spelling

leenipper replied:
ok

petertseng commented on exercises/bowling/example_gen.go:
```diff
@@ -90,14 +90,14 @@ var scoreTestCases = []struct {
 	valid          bool     // true => no error, false => error expected
 	score          int	// when .valid == true, the expected score value
-	errorString    string   // when .valid == false, error string detail
+	explainText    string   // when .valid == false, error explaination text
```
spelling

petertseng commented on exercises/bowling/example_gen.go:
```diff
@@ -106,14 +106,14 @@ var rollTestCases = []struct {
 	valid          bool     // true => no error, false => error expected
 	roll           int	// pin count for the test roll
-	errorString    string   // when .valid == false, error string detail
+	explainText    string   // when .valid == false, error explaination text
```
spelling

leenipper replied:
ok

petertseng commented on exercises/bowling/example_gen.go:
```diff
@@ -0,0 +1,119 @@
+// Template to generate two sets of test cases, one for Score tests and one for Roll tests.
+var tmpl = `package bowling
+
+// Source: {{.Ori}}
```
if rebased on master, these three lines can become `{{.Header}}``

leenipper commented:
Ok, I updated with .Header.
I also added an error test which I stumbled on when injecting errors on previous go-round. The x-common/exercises/bowling/canonical-data.json needs another test case for it.

petertseng commented:
Thanks for the many new exercises you've added to the track!

leenipper commented:
You are welcome!
